<a href="https://colab.research.google.com/github/ckeerthana7/MLC_Lab-Activities/blob/main/DataClean_S.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
# import packages
import pandas as pd

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 1) Load the messy dataset (with headers)

In [3]:
# messy_employee_health.csv
DIR = '/content/drive/My Drive/Colab Notebooks/mlc'
datafile = 'messy_employee_health.csv'
filename = DIR+'/'+datafile

In [4]:
import os
os.chdir(DIR)

In [65]:
df = pd.read_csv(filename)
df

ID  Full Name   Age    GENDER Heart Rate  Height  \
0     1.0       Anu   22.0   Female     72 bpm  170 cm   
1     2.0     Rahul    NaN     MALE      85bpm  1.75 m   
2     3.0     Meera   29.0  female         NaN  175 cm   
3     4.0      Sidd   -5.0     Male     90 bpm   180cm   
4     5.0     Priya   31.0   FEMALE    100 BPM  1.60 m   
5     6.0     Kiran  120.0     male         ??  160 cm   
6     7.0       Anu   22.0   Female     72 bpm  170 cm   
7     8.0      Zara   27.0   FeMale     65 bpm     NaN   
8     9.0       NaN   24.0     male     75 bpm  175 cm   
9    10.0     Vijay   28.0      Mle     80 bpm  170 cm   
10    NaN       NaN    NaN      NaN        NaN     NaN   

                                   Address       Join Date       Notes  \
0            12 MG Road, Bengaluru, 560001      2024-01-05        good   
1              45 Park St, Kolkata, 700016      05/02/2024  naïve café   
2           77 Anna Salai, Chennai, 600002      2024/03/10         NaN   
3              1 Marine Dr, Mumbai, 400001  March 15, 2024           ?   
4   No. 9, Residency Rd, Bengaluru, 560025      2024-13-01         NaN   
5                      Somewhere, , 560002      2024-06-01     unknown   
6            12 MG Road, Bengaluru, 560001      2024-01-05        good   
7           8 Lake View, Hyderabad, 500001      2024-07-04           —   
8          15 Church St, Bengaluru, 560001             NaN          ok   
9               18 North St, Delhi, 110001      2024-09-01          ok   
10                                     NaN             NaN         NaN   

   Constant Column Irrelevant Column  Empty Column  
0             SAME                x1           NaN  
1             SAME                x2           NaN  
2             SAME                x3           NaN  
3             SAME                x4           NaN  
4             SAME                x5           NaN  
5             SAME                x6           NaN  
6             SAME                x1           NaN  
7             SAME                x7           NaN  
8             SAME                x8           NaN  
9             SAME                x9           NaN  
10             NaN               NaN           NaN

In [66]:
df.columns

Index(['  ID ', 'Full Name', 'Age ', 'GENDER', 'Heart Rate', 'Height',
       'Address', 'Join Date', 'Notes', 'Constant Column', 'Irrelevant Column',
       'Empty Column'],
      dtype='object')

In [67]:
#rename the column names
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')

In [68]:
df.columns

Index(['id', 'full_name', 'age', 'gender', 'heart_rate', 'height', 'address',
       'join_date', 'notes', 'constant_column', 'irrelevant_column',
       'empty_column'],
      dtype='object')

In [87]:
df.dtypes

id                            Int64
full_name                    object
age                           Int64
gender                       object
heart_rate           string[python]
height                       object
address                      object
join_date            datetime64[ns]
notes                        object
constant_column              object
irrelevant_column            object
empty_column                float64
dtype: object

In [70]:
#convert the dtypes
df['id'] = df['id'].astype('Int64')
df['age'] = df['age'].astype('Int64')


In [71]:
df['join_date'][4]= '2024-01-01'

/tmp/ipython-input-666666167.py:1: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df['join_date'][4]= '2024-01-01'
/tmp/ipython-input-666666167.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame



In [72]:
df['join_date']= df['join_date'].astype('datetime64[ns]')

In [73]:
df

id full_name   age   gender heart_rate  height  \
0      1       Anu    22   Female     72 bpm  170 cm   
1      2     Rahul  <NA>     MALE      85bpm  1.75 m   
2      3     Meera    29  female         NaN  175 cm   
3      4      Sidd    -5     Male     90 bpm   180cm   
4      5     Priya    31   FEMALE    100 BPM  1.60 m   
5      6     Kiran   120     male         ??  160 cm   
6      7       Anu    22   Female     72 bpm  170 cm   
7      8      Zara    27   FeMale     65 bpm     NaN   
8      9       NaN    24     male     75 bpm  175 cm   
9     10     Vijay    28      Mle     80 bpm  170 cm   
10  <NA>       NaN  <NA>      NaN        NaN     NaN   

                                   address  join_date       notes  \
0            12 MG Road, Bengaluru, 560001 2024-01-05        good   
1              45 Park St, Kolkata, 700016 2024-05-02  naïve café   
2           77 Anna Salai, Chennai, 600002 2024-03-10         NaN   
3              1 Marine Dr, Mumbai, 400001 2024-03-15           ?   
4   No. 9, Residency Rd, Bengaluru, 560025 2024-01-01         NaN   
5                      Somewhere, , 560002 2024-06-01     unknown   
6            12 MG Road, Bengaluru, 560001 2024-01-05        good   
7           8 Lake View, Hyderabad, 500001 2024-07-04           —   
8          15 Church St, Bengaluru, 560001        NaT          ok   
9               18 North St, Delhi, 110001 2024-09-01          ok   
10                                     NaN        NaT         NaN   

   constant_column irrelevant_column  empty_column  
0             SAME                x1           NaN  
1             SAME                x2           NaN  
2             SAME                x3           NaN  
3             SAME                x4           NaN  
4             SAME                x5           NaN  
5             SAME                x6           NaN  
6             SAME                x1           NaN  
7             SAME                x7           NaN  
8             SAME                x8           NaN  
9             SAME                x9           NaN  
10             NaN               NaN           NaN

In [74]:
df['gender'] = df['gender'].astype('category')

In [76]:
df['gender'] = df['gender'].str.lower()

In [82]:
df

id full_name   age   gender heart_rate  height  \
0      1       Anu    22   female     72 bpm  170 cm   
1      2     Rahul  <NA>     male      85bpm  1.75 m   
2      3     Meera    29  female         NaN  175 cm   
3      4      Sidd     5     male     90 bpm   180cm   
4      5     Priya    31   female    100 BPM  1.60 m   
5      6     Kiran   120     male         ??  160 cm   
6      7       Anu    22   female     72 bpm  170 cm   
7      8      Zara    27   female     65 bpm     NaN   
8      9       NaN    24     male     75 bpm  175 cm   
9     10     Vijay    28     male     80 bpm  170 cm   
10  <NA>       NaN  <NA>      NaN        NaN     NaN   

                                   address  join_date       notes  \
0            12 MG Road, Bengaluru, 560001 2024-01-05        good   
1              45 Park St, Kolkata, 700016 2024-05-02  naïve café   
2           77 Anna Salai, Chennai, 600002 2024-03-10         NaN   
3              1 Marine Dr, Mumbai, 400001 2024-03-15           ?   
4   No. 9, Residency Rd, Bengaluru, 560025 2024-01-01         NaN   
5                      Somewhere, , 560002 2024-06-01     unknown   
6            12 MG Road, Bengaluru, 560001 2024-01-05        good   
7           8 Lake View, Hyderabad, 500001 2024-07-04           —   
8          15 Church St, Bengaluru, 560001        NaT          ok   
9               18 North St, Delhi, 110001 2024-09-01          ok   
10                                     NaN        NaT         NaN   

   constant_column irrelevant_column  empty_column  
0             SAME                x1           NaN  
1             SAME                x2           NaN  
2             SAME                x3           NaN  
3             SAME                x4           NaN  
4             SAME                x5           NaN  
5             SAME                x6           NaN  
6             SAME                x1           NaN  
7             SAME                x7           NaN  
8             SAME                x8           NaN  
9             SAME                x9           NaN  
10             NaN               NaN           NaN

In [78]:
df['gender'][9]= 'male'

/tmp/ipython-input-486174259.py:1: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df['gender'][9]= 'male'
/tmp/ipython-input-486174259.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the c

In [81]:
df['age'][3]= 5

In [91]:
df

id full_name   age   gender heart_rate  height  \
0      1       Anu    22   female     72 bpm  170 cm   
1      2     Rahul  <NA>     male      85bpm  1.75 m   
2      3     Meera    29  female        <NA>  175 cm   
3      4      Sidd     5     male     90 bpm   180cm   
4      5     Priya    31   female    100 BPM  1.60 m   
5      6     Kiran   120     male         ??  160 cm   
6      7       Anu    22   female     72 bpm  170 cm   
7      8      Zara    27   female     65 bpm     NaN   
8      9       NaN    24     male     75 bpm  175 cm   
9     10     Vijay    28     male     80 bpm  170 cm   
10  <NA>       NaN  <NA>      NaN       <NA>     NaN   

                                   address  join_date       notes  \
0            12 MG Road, Bengaluru, 560001 2024-01-05        good   
1              45 Park St, Kolkata, 700016 2024-05-02  naïve café   
2           77 Anna Salai, Chennai, 600002 2024-03-10         NaN   
3              1 Marine Dr, Mumbai, 400001 2024-03-15           ?   
4   No. 9, Residency Rd, Bengaluru, 560025 2024-01-01         NaN   
5                      Somewhere, , 560002 2024-06-01     unknown   
6            12 MG Road, Bengaluru, 560001 2024-01-05        good   
7           8 Lake View, Hyderabad, 500001 2024-07-04           —   
8          15 Church St, Bengaluru, 560001        NaT          ok   
9               18 North St, Delhi, 110001 2024-09-01          ok   
10                                     NaN        NaT         NaN   

   constant_column irrelevant_column  empty_column  
0             SAME                x1           NaN  
1             SAME                x2           NaN  
2             SAME                x3           NaN  
3             SAME                x4           NaN  
4             SAME                x5           NaN  
5             SAME                x6           NaN  
6             SAME                x1           NaN  
7             SAME                x7           NaN  
8             SAME                x8           NaN  
9             SAME                x9           NaN  
10             NaN               NaN           NaN

In [92]:
#for row
df= df.dropna(axis=0, how="all")

In [94]:
df.shape

(10, 12)

In [95]:
#column
df= df.dropna(axis=1, how="all")

In [96]:
df.shape

(10, 11)

In [97]:
df

id full_name   age   gender heart_rate  height  \
0   1       Anu    22   female     72 bpm  170 cm   
1   2     Rahul  <NA>     male      85bpm  1.75 m   
2   3     Meera    29  female        <NA>  175 cm   
3   4      Sidd     5     male     90 bpm   180cm   
4   5     Priya    31   female    100 BPM  1.60 m   
5   6     Kiran   120     male         ??  160 cm   
6   7       Anu    22   female     72 bpm  170 cm   
7   8      Zara    27   female     65 bpm     NaN   
8   9       NaN    24     male     75 bpm  175 cm   
9  10     Vijay    28     male     80 bpm  170 cm   

                                  address  join_date       notes  \
0           12 MG Road, Bengaluru, 560001 2024-01-05        good   
1             45 Park St, Kolkata, 700016 2024-05-02  naïve café   
2          77 Anna Salai, Chennai, 600002 2024-03-10         NaN   
3             1 Marine Dr, Mumbai, 400001 2024-03-15           ?   
4  No. 9, Residency Rd, Bengaluru, 560025 2024-01-01         NaN   
5                     Somewhere, , 560002 2024-06-01     unknown   
6           12 MG Road, Bengaluru, 560001 2024-01-05        good   
7          8 Lake View, Hyderabad, 500001 2024-07-04           —   
8         15 Church St, Bengaluru, 560001        NaT          ok   
9              18 North St, Delhi, 110001 2024-09-01          ok   

  constant_column irrelevant_column  
0            SAME                x1  
1            SAME                x2  
2            SAME                x3  
3            SAME                x4  
4            SAME                x5  
5            SAME                x6  
6            SAME                x1  
7            SAME                x7  
8            SAME                x8  
9            SAME                x9

In [99]:
#divide the address into different rows
# n=2 means it will split at most twice, resulting in 3 columns
df[['street', 'city', 'pincode']] = df['address'].str.split(',', n=2, expand=True)

In [100]:
df

id full_name   age   gender heart_rate  height  \
0   1       Anu    22   female     72 bpm  170 cm   
1   2     Rahul  <NA>     male      85bpm  1.75 m   
2   3     Meera    29  female        <NA>  175 cm   
3   4      Sidd     5     male     90 bpm   180cm   
4   5     Priya    31   female    100 BPM  1.60 m   
5   6     Kiran   120     male         ??  160 cm   
6   7       Anu    22   female     72 bpm  170 cm   
7   8      Zara    27   female     65 bpm     NaN   
8   9       NaN    24     male     75 bpm  175 cm   
9  10     Vijay    28     male     80 bpm  170 cm   

                                  address  join_date       notes  \
0           12 MG Road, Bengaluru, 560001 2024-01-05        good   
1             45 Park St, Kolkata, 700016 2024-05-02  naïve café   
2          77 Anna Salai, Chennai, 600002 2024-03-10         NaN   
3             1 Marine Dr, Mumbai, 400001 2024-03-15           ?   
4  No. 9, Residency Rd, Bengaluru, 560025 2024-01-01         NaN   
5                     Somewhere, , 560002 2024-06-01     unknown   
6           12 MG Road, Bengaluru, 560001 2024-01-05        good   
7          8 Lake View, Hyderabad, 500001 2024-07-04           —   
8         15 Church St, Bengaluru, 560001        NaT          ok   
9              18 North St, Delhi, 110001 2024-09-01          ok   

  constant_column irrelevant_column         street           city  \
0            SAME                x1     12 MG Road      Bengaluru   
1            SAME                x2     45 Park St        Kolkata   
2            SAME                x3  77 Anna Salai        Chennai   
3            SAME                x4    1 Marine Dr         Mumbai   
4            SAME                x5          No. 9   Residency Rd   
5            SAME                x6      Somewhere                  
6            SAME                x1     12 MG Road      Bengaluru   
7            SAME                x7    8 Lake View      Hyderabad   
8            SAME                x8   15 Church St      Bengaluru   
9            SAME                x9    18 North St          Delhi   

              pincode  
0              560001  
1              700016  
2              600002  
3              400001  
4   Bengaluru, 560025  
5              560002  
6              560001  
7              500001  
8              560001  
9              110001

In [101]:
df=df.drop(columns=['address'])

In [102]:
df

id full_name   age   gender heart_rate  height  join_date       notes  \
0   1       Anu    22   female     72 bpm  170 cm 2024-01-05        good   
1   2     Rahul  <NA>     male      85bpm  1.75 m 2024-05-02  naïve café   
2   3     Meera    29  female        <NA>  175 cm 2024-03-10         NaN   
3   4      Sidd     5     male     90 bpm   180cm 2024-03-15           ?   
4   5     Priya    31   female    100 BPM  1.60 m 2024-01-01         NaN   
5   6     Kiran   120     male         ??  160 cm 2024-06-01     unknown   
6   7       Anu    22   female     72 bpm  170 cm 2024-01-05        good   
7   8      Zara    27   female     65 bpm     NaN 2024-07-04           —   
8   9       NaN    24     male     75 bpm  175 cm        NaT          ok   
9  10     Vijay    28     male     80 bpm  170 cm 2024-09-01          ok   

  constant_column irrelevant_column         street           city  \
0            SAME                x1     12 MG Road      Bengaluru   
1            SAME                x2     45 Park St        Kolkata   
2            SAME                x3  77 Anna Salai        Chennai   
3            SAME                x4    1 Marine Dr         Mumbai   
4            SAME                x5          No. 9   Residency Rd   
5            SAME                x6      Somewhere                  
6            SAME                x1     12 MG Road      Bengaluru   
7            SAME                x7    8 Lake View      Hyderabad   
8            SAME                x8   15 Church St      Bengaluru   
9            SAME                x9    18 North St          Delhi   

              pincode  
0              560001  
1              700016  
2              600002  
3              400001  
4   Bengaluru, 560025  
5              560002  
6              560001  
7              500001  
8              560001  
9              110001

# Investigate
*  null values
*  column names has space
*  special characters
*  case
*  data types
*  join date type is object
*  address values
*  1 column is empty



## 2) Missing headers (use the no-header file)


In [8]:
#remove the header
df1 = pd.read_csv(filename, header=None)
df1

0          1     2        3           4       5   \
0     ID   Full Name  Age    GENDER  Heart Rate  Height   
1       1        Anu    22   Female      72 bpm  170 cm   
2       2      Rahul   NaN     MALE       85bpm  1.75 m   
3       3      Meera    29  female          NaN  175 cm   
4       4       Sidd    -5     Male      90 bpm   180cm   
5       5      Priya    31   FEMALE     100 BPM  1.60 m   
6       6      Kiran   120     male          ??  160 cm   
7       7        Anu    22   Female      72 bpm  170 cm   
8       8       Zara    27   FeMale      65 bpm     NaN   
9       9        NaN    24     male      75 bpm  175 cm   
10     10      Vijay    28      Mle      80 bpm  170 cm   
11    NaN        NaN   NaN      NaN         NaN     NaN   

                                        6               7           8   \
0                                  Address       Join Date       Notes   
1            12 MG Road, Bengaluru, 560001      2024-01-05        good   
2              45 Park St, Kolkata, 700016      05/02/2024  naïve café   
3           77 Anna Salai, Chennai, 600002      2024/03/10         NaN   
4              1 Marine Dr, Mumbai, 400001  March 15, 2024           ?   
5   No. 9, Residency Rd, Bengaluru, 560025      2024-13-01         NaN   
6                      Somewhere, , 560002      2024-06-01     unknown   
7            12 MG Road, Bengaluru, 560001      2024-01-05        good   
8           8 Lake View, Hyderabad, 500001      2024-07-04           —   
9          15 Church St, Bengaluru, 560001             NaN          ok   
10              18 North St, Delhi, 110001      2024-09-01          ok   
11                                     NaN             NaN         NaN   

                 9                  10            11  
0   Constant Column  Irrelevant Column  Empty Column  
1              SAME                 x1           NaN  
2              SAME                 x2           NaN  
3              SAME                 x3           NaN  
4              SAME                 x4           NaN  
5              SAME                 x5           NaN  
6              SAME                 x6           NaN  
7              SAME                 x1           NaN  
8              SAME                 x7           NaN  
9              SAME                 x8           NaN  
10             SAME                 x9           NaN  
11              NaN                NaN           NaN

In [ ]:
# check the first first records


## 3) Missing values
Compute missing counts and percentages. Then choose a strategy (drop, fill, impute).

In [ ]:
# drop the missing values


# fill the missing values with average


# impute the values

## 4) Empty rows and empty columns
Identify and remove completely empty rows and columns.

In [ ]:
# TODO: find empty rows

# TODO: drop empty rows


# TODO: find empty columns


# TODO: drop empty columns

# df_step.shape

## 5) Messy column names
Standardize: strip, lowercase, underscores, remove special chars.

In [ ]:
# TODO: clean column names




## 6) Multiple variables in one column (Heart Rate)
Extract numeric heart rate (bpm) as an integer. Handle values like `85bpm`, `100 BPM`, `??`.

In [ ]:
# TODO: normalize heart_rate strings then extract digits
# df_step['heart_rate'] = ...
# df_step['heart_rate_bpm'] = ...

# df_step[['heart_rate', 'heart_rate_bpm']].head(10)

## 7) Inconsistent units (Height)
Convert height to meters (float). Handle `170 cm`, `1.75 m`, `180cm`.

In [ ]:
# TODO: create height_m
# df_step['height_m'] = ...

# df_step[['height', 'height_m']].head(10)

## 8) Capitalization + typos in categorical values (Gender)
Standardize gender values and fix typos like `Mle`.

In [ ]:
# TODO: normalize gender
# df_step['gender'] = ...

# df_step['gender'].value_counts(dropna=False)

## 9) Data types (Age, IDs)
Convert `id` to integer (nullable ok), `age` to numeric; identify invalid values (e.g., negative age).

In [ ]:
# TODO: convert types
# df_step['id'] = ...
# df_step['age'] = ...

# TODO: find invalid ages
# df_step[(df_step['age'] < 0) | (df_step['age'] > 100)]

## 10) Addresses
Split address into street, city, and pincode. Handle missing parts safely.

In [ ]:
# TODO: split address by comma into 3 parts
# addr = df_step['address'].str.split(',', n=2, expand=True)
# df_step['street'] = ...
# df_step['city'] = ...
# df_step['pincode'] = ...

# df_step[['address','street','city','pincode']].head(10)

## 11) Dates
Parse join_date with mixed formats and coerce invalid dates to NaT.

In [ ]:
# TODO: parse dates
# df_step['join_date'] = pd.to_datetime(df_step['join_date'], errors='coerce', dayfirst=True)
# df_step[['join_date']].head(10)

## 12) Non-ASCII characters (Notes)
Remove non-ASCII characters OR normalize them.

In [ ]:
# TODO: remove non-ascii
# df_step['notes_clean'] = ...
# df_step[['notes','notes_clean']].head(10)

## 13) Unnecessary data
Remove duplicate rows, irrelevant columns, and uninformative columns.

In [ ]:
# TODO: duplicates
# df_step.duplicated().sum()
# df_step = df_step.drop_duplicates()

# TODO: drop irrelevant columns
# df_step = df_step.drop(columns=['irrelevant_column'], errors='ignore')

# TODO: drop uninformative columns (nunique == 1)
# nun = df_step.nunique(dropna=False)
# drop_cols = nun[nun == 1].index
# df_step = df_step.drop(columns=drop_cols)

# df_step.shape

## 14) Final cleaned dataset
Show summary and export as CSV.